File: LeetCode_Java/src/com/leetcode/rp/LeetCode13.java
package com.leetcode.rp;

import java.util.ArrayList;
import java.util.HashMap;

class Solution {
    public int romanToInt(String s) {
        int ans = 0;
        HashMap<Character, Integer> hashMap = new HashMap<>();
        hashMap.put('I', 1);
        hashMap.put('V', 5);
        hashMap.put('X', 10);
        hashMap.put('L', 50);
        hashMap.put('C', 100);
        hashMap.put('D', 500);
        hashMap.put('M', 1000);
        for (int i = 0; i < s.length() - 1; i++) {
            if (hashMap.get(s.charAt(i)) < hashMap.get(s.charAt(i + 1)))
                ans -= hashMap.get(s.charAt(i));
            else
                ans += hashMap.get(s.charAt(i));
        }
        return ans + hashMap.get(s.charAt(s.length() - 1));
    }
}

public class LeetCode13 {
    public static void main(String[] args) {
        Solution solution = new Solution();
        ArrayList List = new ArrayList();
        List.add("III");
        List.add("IV");
        List.add("IX");
        List.add("LVIII");
        List.add("MCMXCIV");
        for (Object elem : List) {
            System.out.println(elem.toString());
            System.out.println(solution.romanToInt(elem.toString()));
        }
    }
}

File: LeetCode_Scala/src/com/rp/leetcode/LeetCode13.scala
package com.rp.leetcode

import scala.collection.mutable

object LeetCode13 {
  def romanToInt(s: String): Int = {
    var ans = 0
    val map = new mutable.HashMap[Char, Int]()
    map.put('I', 1)
    map.put('V', 5)
    map.put('X', 10)
    map.put('L', 50)
    map.put('C', 100)
    map.put('D', 500)
    map.put('M', 1000)
    for (i <- 0 to (s.length - 2)) {
      if (map(s.charAt(i + 1)) > map(s.charAt(i))) {
        ans -= map(s.charAt(i))
      }
      else {
        ans += map(s.charAt(i))
      }
    }
    ans + map(s.charAt(s.length - 1))
  }
  val m = Map('I' -> 1, 'V' -> 5, 'X' -> 10, 'L' -> 50, 'C' -> 100, 'D' -> 500, 'M' -> 1000)
  def romanToIntImproved(s: String): Int = {
    (s zip s.tail).foldLeft(0) {
      case ((sum, (ch1, ch2))) =>
        if (m(ch1) < m(ch2))
          sum - m(ch1)
        else
          sum + m(ch1)
    } + m(s.last)
  }

  def main(args: Array[String]): Unit = {
    val samples = Array("III", "IV", "IX", "LVIII", "MCMXCIV")
    for(i <- 0 to (samples.length - 1)){
      println(romanToInt(samples(i)))
    }
  }
}
